Promise (Lazy) Support + Global onActivation/onDeactivation

parisholley opened the pull request:
This is a cleaner refactor which combines #1074 and #1086 to allow for users to opt-in to lazy loaded dependencies by using an asynchronous API.

For anyone wanting to play around with this, you can use the NPM release:

`@parisholley/inversify-async:1.1.7`

## Description

New container methods for pulling objects out of the container in an asynchronous fashion while respecting container semantics and error if trying to mix synchronous/asynchronous dependencies.

**Implementation**
```javascript
container.bind<Type>('name').toDynamicValue(async () => {
  return await myAsyncFunction();
});
container.bind<string>('resolve').toDynamicValue(() => Promise.resolve('foobar'));

const value = await container.getAsync<Type>('name'); // result of myAsyncFunction
const value = await container.get<Type>('name'); // throws an exception
const value = await container.getAsync<string>('resolve'); // foobar
````
**Global onActivation**

```javascript
container.onActivation('asyncIdentifier', (context, injectable) => {
  return Promise.resolve('newObject'); // returning a promise enforces use of getAsync(), even if future binds are not async
});
container.bind('asyncIdentifier').toConstantValue('foobar');

await container.getAsync('asyncIdentifier');

container.onActivation('syncIdentifier', (context, injectable) => {
  return 'newObject'; // no promise = use legacy API
});

container.get('syncIdentifier');
````

**Binding onActivation**

```javascript
container.bind('asyncIdentifier').toConstantValue('foobar').onActivation((context, injectable) => {
  return Promise.resolve('newObject'); // returning a promise enforces use of getAsync()
});

await container.getAsync('asyncIdentifier');

container.bind('asyncIdentifier').toConstantValue('foobar').onActivation((context, injectable) => {
  return 'newObject'; // no promise = use legacy API
});

container.get('syncIdentifier');
````

**Parent/Child onActivation**

```javascript
container.onActivation('asyncIdentifier', (context, injectable) => {
  return Promise.resolve('newObject'); // returning a promise enforces use of getAsync()
});

const child = container.createChild();
child.bind('asyncIdentifier').toConstantValue('foobar');

await container.getAsync('asyncIdentifier');
````

**Module onActivation**

```javascript
container.bind('asyncIdentifier').toConstantValue('foobar');

const TestContainerModule = new ContainerModule((bind, unbind, isBound, rebind, onActivation) => {
  onActivation('asyncIdentifier', async (context, service) => {

 }));
});
````

## Related Issue

#418 
#475
#887
#404

## Motivation and Context

**Use Case 1: Unit/System Testing**
While you technically could create separate containers (with a parent container for common objects) for a unit test environment and system test (database, etc), it adds more code complexity and also requires each system integration be live and working in order to bootstrap the container (this matters, because for instance in the IDE, if I want to run a single test, with a single I/O dependency, I don't have to initiate every connection).

**Use Case 2: Common Container Use**
Let's say you have a "service" code base that is re-used between a web application, command line interface and batch processing. In the case of the CLI, you have different commands that serve different purpose, eg:
- Clone production data
- Invoke third party service

By making I/O dependencies lazy loaded (async), our CLI can wait until runtime to connect to the system that it needs for the task (otherwise, same problem as above, you are forced to connect-to, and have access-to systems that are unrelated to the command at hand).

**Use Case 3: Transactions:**
For testing (or perhaps a custom request based framework where child containers are created for each web request to implement "Request Scope" similar to spring), it may be useful to have async construct/destroy methods. For example, a web request where we want to run all service calls in a transaction and commit on at the end (by unbinding the container), or in tests by rolling back database changes for more performant runs.

## How Has This Been Tested?

Full suite of tests have been implemented.

## Types of changes

- [x] Updated docs / Refactor code / Added a tests case (non-breaking change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have updated the changelog.

greenkeeper[bot] committed:
chore(package): update publish-please to version 5.5.0

parisholley committed:
potential appveyor fix

parisholley committed:
fix for Microsoft/TSJS-lib-generator#559

parisholley committed:
fix for Microsoft/TSJS-lib-generator#559

parisholley committed:
Merge branch 'appveyor' into TimerHandlerFix

parisholley committed:
full promise/async support, global onActivation/onDeactivation, binding onDeactivation

parisholley committed:
appveyor works again?

parisholley committed:
latest versions of FF require latest launcher to not hang

parisholley committed:
latest versions of FF require latest launcher to not hang

parisholley committed:
Merge branch 'appveyor-fix' into async-v2

parisholley committed:
context bug

parisholley committed:
merged with master

parisholley committed:
merged with latest

sublimator commented on package.json:
```diff
@@ -75,7 +75,7 @@
     "sinon": "7.4.1",
     "tslint": "5.19.0",
-    "typescript": "^3.5.3",
+    "typescript": "3.0.3",
```
Why the version drop ?

parisholley replied:
https://github.com/inversify/InversifyJS/pull/1130

sublimator replied:
thanks

sublimator commented on src/resolution/instantiation.ts:
```diff
@@ -33,39 +32,70 @@ function _createInstance(Func: interfaces.Newable<any>, injections: Object[]): a
 
         const constructorInjections = constructorInjectionsRequests.map(resolveRequest);
 
+        if (constructorInjections.some((ci) => ci instanceof Promise)) {
```
This isn't that massive a change huh ?

parisholley replied:
nope :) i hacked up half the project the first time around trying to get this to work, and then realized it would be as simple as promise detection.

sublimator replied:
Ponder points:
1.  3rd party Promise libs. Do they generally just inherit from Promise ? I never use them, but I guess some people (still?) do.
2. What happens when you have "concurrent" async requests going on at once for a given container (Say a parent container, and per http request child containers accessing the parent container for 'shared' deps). Is that supported ? Does this do the right thing in terms of honoring the scopes etc ?

parisholley replied:
1. i'd assume if anyone is using a 3rd party lib, they are probably doing so as a polyfill, which would override window.Promise, and this should still be true.

2. from what I understand of my latest implementation, if there is a race condition with async, it would also exist with non-async (eg: dynamic values, etc). the same semantics of how things are looked up in parent containers and/or cached on bindings ring true with promises.

sublimator replied:
I see, thanks :) So the promises themselves are cached if I am understanding correctly?

parisholley committed:
support promise retying

parisholley committed:
change promise checks to thenable checks

parisholley committed:
fix dangling promise

sublimator commented on src/utils/async.ts:
```diff
@@ -0,0 +1,5 @@
+function isPromise(object: any): boolean {
```
can change
```
function isPromise(object: any): boolean {
```
to
```
function isPromise(obj: any): obj is Promise {
```

then you will get auto type narrowing, so inside if blocks where it's true, no need to cast

parisholley replied:
learn something everyday :)

sublimator replied:
Good stuff coming in 3.7 re: type narrowing :)
https://github.com/microsoft/TypeScript/pull/32695

parisholley committed:
take advantage of type narrowing

parisholley committed:
Merge branch 'master' into async-v2

parisholley committed:
export predestroy

parisholley committed:
fix when promises are rejected and null values end up in the cache

dcavanagh committed:
Merge branch 'master' into async-v2

PodaruDragos committed:
Merge remote-tracking branch 'origin' into pr/parisholley/1132

PodaruDragos committed:
fix failing ci

dependabot[bot] committed:
Bump elliptic from 6.5.3 to 6.5.4 (#1298)

Bumps [elliptic](https://github.com/indutny/elliptic) from 6.5.3 to 6.5.4.
- [Release notes](https://github.com/indutny/elliptic/releases)
- [Commits](https://github.com/indutny/elliptic/compare/v6.5.3...v6.5.4)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

tonyhallett committed:
Apply middleware when Container.resolve - fixes 1128 (#1129)

* middleware applied

tonyhallett committed:
Setter injection without cast (#1152)

* setter injection without cast

notaphplover committed:
test(resolution): remove additional spaces

notaphplover committed:
feat: refactor Container extracting logic to  _preDestroyBinding and _removeServiceFromDictionary

notaphplover committed:
refactor: updated several modules to pass codeclimate's analysis.

This is a squash commit. It includes the following changes:

style: remove trailing whitespace
test: add async deactivation test case
feat: update container.doDeactivation to use a simpler iterableiterator
test: update test to be async
refactor: extract common decorator logic to propertyEventDecorator
refactor: update Container._doDeactivation method and calls to remove redundant error handling logic
refactor: update Container._doDeactivation to get instance constructor in the same way as in the rest of the project
style: update tsConfig with missing line feed
refactor: update _doServiceBindingsDeactivation to not to receive a binding
refactor: update activationLoop to receive an iterable iterator of activation bindings
refactor: update _onActivation with less paramenters and loc.
refactor: update onActivation extracting container iterator generation in a _onActivationGetContainersToTraverse method
refactor: update _onActivation to avoid recursion. An iterative strategy has been used instead of a recursive one.
refactor:_activationLoop no longer needs more than 4 parameters and it's logic has been extracted to several methods
update onActivation to reduce cognitive complexity
refactor: update resolveInstance to reduce its cognitive complexity
refactor: extract common logic to _filterRequestsByTargetType method
refactor: refactor _get to reduce paramenters and split async handling in a _getWithAsyncContext method
refactor: express AsyncContainerModuleCallBack as an AsyncCallback of ContainerModuleCallBack
style: add missing line feed

notaphplover committed:
constant value / function binding activated when resolved (#1122)

notaphplover committed:
Merge branch 'master' of https://github.com/inversify/InversifyJS into async-v2

notaphplover committed:
Merge remote-tracking branch 'upstream/master' into async-v2

notaphplover commented on src/container/lookup.ts:
```diff
@@ -1,7 +1,7 @@
 import { interfaces } from "../interfaces/interfaces";
 
-class Lookup<T extends interfaces.Clonable<T>> implements interfaces.Lookup<T> {
+class Lookup<T extends interfaces.Clonable<any> | any> implements interfaces.Lookup<T> {
```
Maybe we can just remove the constraint:

```suggestion
class Lookup<T> implements interfaces.Lookup<T> {
```

PodaruDragos replied:
agree

notaphplover commented on src/container/lookup.ts:
```diff
@@ -91,7 +91,7 @@ class Lookup<T extends interfaces.Clonable<T>> implements interfaces.Lookup<T> {
 
         this._map.forEach((value, key) => {
-            value.forEach((b) => copy.add(key, b.clone()));
+            value.forEach((b: any) => copy.add(key, b.clone ? b.clone() : b));
```
I would create a Typeguard instead:

**isClonable.ts**
```Typescript
   function isClonable<T>(obj: unknown): obj is interfaces.Clonable<T> {
        return (typeof obj === 'object') 
            && (obj !== null)
            && ('clone' in obj)
            && typeof (obj as interfaces.Clonable<T>).clone === 'function';
    }
```

```Typescript
this._map.forEach((value, key) => {
    value.forEach((b: any) => copy.add(key, isClonable<T>(b) ? b.clone() : b));
});
```

In addition, maybe we should not clone a non clonable object but throw an exception instead

PodaruDragos replied:
also agree with the typeguard

notaphplover reviewed (commented):
Just added some comments

PodaruDragos commented on src/constants/error_msgs.ts:
```diff
@@ -17,6 +17,8 @@ export const INVALID_BINDING_TYPE = "Invalid binding type:";
 export const NO_MORE_SNAPSHOTS_AVAILABLE = "No snapshot available to restore.";
 export const INVALID_MIDDLEWARE_RETURN = "Invalid return type in middleware. Middleware must return!";
 export const INVALID_FUNCTION_BINDING = "Value provided to function binding must be a function!";
+export const LAZY_IN_SYNC = (key: any) => `You are attempting to construct '${key}' in a synchronous way
```
can we maybe replace any's with unknowns ?

PodaruDragos commented on src/interfaces/interfaces.ts:
```diff
@@ -1,5 +1,9 @@
 
+    type AsyncCallback<TCallback> =
+        TCallback extends (...args: infer TArgs) => infer TResult ? (...args: TArgs) => Promise<TResult>
+        : never;
```
nice !

PodaruDragos commented on src/interfaces/interfaces.ts:
```diff
@@ -176,6 +185,14 @@ namespace interfaces {
         getAllNamed<T>(serviceIdentifier: ServiceIdentifier<T>, named: string | number | symbol): T[];
+        getAsync<T>(serviceIdentifier: ServiceIdentifier<T>): Promise<T>;
+        getNamedAsync<T>(serviceIdentifier: ServiceIdentifier<T>, named: string | number | symbol): Promise<T>;
+        getTaggedAsync<T>(serviceIdentifier: ServiceIdentifier<T>, key: string | number | symbol, value: any): Promise<T>;
```
we should really figure out a way to do a generic for this value and then apply it to all this kinds of methods.
one suggestion will be to either make it as an unknown or figure out all the types that might come into this and just  type it as a union of all the types

notaphplover replied:
I think value can receive any type, so maybe `unknown` is the best type to apply

PodaruDragos commented on src/resolution/instantiation.ts:
```diff
@@ -29,45 +25,121 @@ function _injectProperties(
+    constr: interfaces.Newable<any>,
+    childRequests: interfaces.Request[],
+    resolveRequest: interfaces.ResolveRequestHandler,
+): any {
```
if we manage to type this, all the work forward to replace any's will be so easy 💯

notaphplover committed:
refactor: update error messages to not to receive any type parameters

notaphplover committed:
refactor: update instantiation with better typings

notaphplover committed:
refactor: update Lookup.clone to use a Typeguard

notaphplover committed:
refactor: update error msg with non any parameters

notaphplover committed:
Merge branch 'master' of https://github.com/inversify/InversifyJS into async-v2

notaphplover committed:
fix: update onActivation test to use a binding with a transient scope instead of a singleton scope one

notaphplover committed:
Merge branch 'master' of https://github.com/inversify/InversifyJS into async-v2

PodaruDragos reviewed (approved):
I went threw this a couple of times ( lol )
LGTM.

@dcavanagh @notaphplover do you guys think we can merge this ?

notaphplover committed:
feat: update container._propagateDeactivation to not to _triggerOnDeactivationAndDestroyMetadata multiple times on an async context

notaphplover committed:
style: make it name shorter

notaphplover committed:
docs: added get async docs

notaphplover commented on src/interfaces/interfaces.ts:
```diff
@@ -42,6 +46,10 @@ namespace interfaces {
         clone(): T;
     }
 
+    export type BindingActivation<T> = (context: interfaces.Context, injectable: T) => T | Promise<T>;
```
Just wondering, should we allow users to work directly with the context? There's no easy way to use it.

It gives access to the current resolution plan which is, at this point, partially resolved. It gives access to the request, which can be used to access all the children and parent requests. The update of these models could lead to an undefined behavior. For example, a user could technically create an infinite loop with a plan cycle. 

I think it's great to provide users a **read only** tool to traverse the dependency tree, but I wonder if `context` is the right tool for them.

What are your thoughts about this? I'd love to have this awesome feature merged, but I prefer to discuss this topic to be sure we provide a great tool with no side effects.

notaphplover replied:
Never mind, onActivation handlers already receive a context

notaphplover committed:
docs: update container docs

notaphplover committed:
docs: add activation handler and pre destroy docs

notaphplover committed:
add missing types and refactor types to use BindingActivation and BindingDeactivation

notaphplover committed:
add deactivation_handler docs

tonyhallett committed:
fix async multi inject and property inject

notaphplover commented on src/container/container.ts:
```diff
@@ -13,6 +13,8 @@ import { getServiceIdentifierAsString } from "../utils/serialization";
 import { ContainerSnapshot } from "./container_snapshot";
 import { Lookup } from "./lookup";
 
+type GetArgs = Omit<interfaces.NextArgs,'contextInterceptor'|'targetType'>
```
I prefer not to use Omit

tonyhallett replied:
Why not ? It is what we are doing by supplying the ommited properties in the _get method

notaphplover replied:
It's true, maybe I'm being too picky. I personally prefer to create a model system which creates types from scratch or extending existing types. But its more a preference rather than an argument, a remanent of my OOP background, we don't need to be bounded by that

notaphplover commented on src/resolution/instantiation.ts:
```diff
@@ -33,17 +13,39 @@ function _createInstance<T>(
+                            isAsync = true
+                            break
+                        }
+                    }
```
This is equivalent to:

```ts
   isAsync = injection.some(isPromise);
```

tonyhallett replied:
Will change

notaphplover commented on src/resolution/instantiation.ts:
```diff
@@ -33,17 +13,39 @@ function _createInstance<T>(
+                    }
+                }else if(isPromise(injection)){
+                    isAsync = true
+                }
```
I think we should try to extract this logic to a method, otherwise the function is too big

tonyhallett replied:
Please feel free - it did pass code climate though.  There are other code climate warnings if you are interested !

notaphplover reviewed (commented):
Great! I just added a couple of comments so far, I think it's an excellent job, I really like those names the GetArgs refactor. Thanks you so much for the fix :)

notaphplover commented on src/resolution/resolver.ts:
```diff
@@ -290,7 +290,7 @@ const _getParentContainersIterator = (container: interfaces.Container, includeSe
 }
 
-function resolve<T>(context: interfaces.Context): T {
+function resolve<T>(context: interfaces.Context): (T | T[] | Promise<T> | Promise<T>[]) {
```
The right return type is `T | Promise<T> | (T | Promise<T>)[]`

notaphplover replied:
Don't worry, I got it, can I push it? This way I can add more typings :)

The difference is subtle but important imo

tonyhallett replied:
will do !

tonyhallett replied:
I will push in a minute or two then is all yours ? Too tired ! Yes the difference is important, in particular to the your review below.

notaphplover replied:
All right, I'll add the changes now or tomorrow morning :) . I hope I'm not being rude. Sometimes it's hard for me to notice

tonyhallett committed:
Container get all async methods to Promise<T[]>. singular async to async functions

notaphplover commented on src/container/container.ts:
```diff
@@ -502,69 +500,68 @@ class Container implements interfaces.Container {
 
-        if (isPromise(result) && !isAsync) {
-            throw new Error(ERROR_MSGS.LAZY_IN_SYNC(defaultArgs.serviceIdentifier));
+        if (isPromise(result)) {
```
if result is `(T | Promise<T>)[]` we could miss the throw, I know how to recreate this and fix it :)

tonyhallett replied:
Spotted in - .some test again

tonyhallett committed:
container resolution type correction and improve lazy in sync test

notaphplover committed:
add typings at resolver and compact boolean logic at container._getButThrowIfAsync

notaphplover committed:
add isPromiseOrContainsPromise

notaphplover committed:
update isPromise to receive unknown and a generic type

tonyhallett committed:
refactor _createInstance

tonyhallett committed:
refactor _resolveRequests with array reduce

notaphplover commented on src/resolution/instantiation.ts:
```diff
@@ -11,35 +11,28 @@ interface InstanceCreationInstruction{
-            isAsync = isPromiseOrContainsPromise(injection);
+        if(!resolvedRequests.isAsync){
+            resolvedRequests.isAsync = isPromiseOrContainsPromise(injection);
         }
```
Maybe we could extract a _resolveRequest method

tonyhallett replied:
I don't think it is necessary now it has been reduced.

notaphplover replied:
true :)

notaphplover commented on src/resolution/instantiation.ts:
```diff
@@ -11,35 +11,28 @@ interface InstanceCreationInstruction{
 }
 
-type CreateInstanceWithInjectionArg<T> = InstanceCreationInstruction & {constr: interfaces.Newable<T>,}
+type ResolvedRequests = InstanceCreationInstruction & {isAsync:boolean}
```
I think we can express this as an interface
```ts
interface ResolvedRequests extends InstanceCreationInstruction { 
    isAsync:boolean;
}
```

tonyhallett replied:
Sure if you prefer

notaphplover commented on src/resolution/instantiation.ts:
```diff
@@ -11,35 +11,28 @@ interface InstanceCreationInstruction{
-type CreateInstanceWithInjectionArg<T> = InstanceCreationInstruction & {constr: interfaces.Newable<T>,}
+type ResolvedRequests = InstanceCreationInstruction & {isAsync:boolean}
+
+type CreateInstanceWithInjectionArg<T> = InstanceCreationInstruction & {constr: interfaces.Newable<T>}
```
```ts
interface CreateInstanceWithInjectionArg<T> extends InstanceCreationInstruction {
    constr: interfaces.Newable<T>;
}
```

tonyhallett replied:
Sure if you prefer

tonyhallett committed:
type to interface

tonyhallett committed:
test coverage

tonyhallett committed:
correct spelling in tests and correct expect in handler

tonyhallett committed:
deactivate refactor and name changes

tonyhallett committed:
refactor activation and deactivation

tonyhallett committed:
update wikis and dynamic value typing

tonyhallett committed:
rebindAsync, missing module functionality and wiki updates

tonyhallett commented on src/bindings/binding.ts:
```diff
@@ -5,7 +5,7 @@ import { id } from "../utils/id";
 class Binding<T> implements interfaces.Binding<T> {
 
     public id: number;
-    public moduleId: string;
```
The value comes from the ContainerModule/AsyncContainerModule id - a number

tonyhallett commented on src/container/container.ts:
```diff
@@ -260,7 +278,8 @@ class Container implements interfaces.Container {
           this._activations.clone(),
-          this._deactivations.clone()
+          this._deactivations.clone(),
+          this._moduleActivationStore.clone()
```
I could not think of a test quickly for the cloning of the moduleActivationStore.

notaphplover replied:
A cloned container should call the handler, so we can check a deactivation handler is called once a cloned container receives an unbind request. That could be a way to test it

tonyhallett replied:
What I was looking for is a test that would fail if I did not clone.  I don't think there is one.  The store for the module handlers is just for removing from deactivations and activations.  Cloning ensures we do not hold on to a handler from a module added after snapshot when we restore.  If we did not clone there would not be a bug just inefficient code.
I will just add a test that the snapshot gets a clone.

tonyhallett commented on src/container/container.ts:
```diff
@@ -459,45 +492,59 @@ class Container implements interfaces.Container {
-        const getRebindFunction = (moduleId: number) =>
+        const getRebindFunction = (moduleId: interfaces.ContainerModuleBase["id"]) =>
             (serviceIdentifier: interfaces.ServiceIdentifier<any>) => {
-                const _rebind = this.rebind.bind(this);
```
I have removed unnecessary js bind from a few places.

tonyhallett commented on wiki/container_api.md:
````diff
@@ -110,7 +110,7 @@ container.bind<Weapon>("Weapon").to(Katana);
 ```
 
-## container.getAsync<T>(serviceIdentifier: ServiceIdentifier<T>)
+## container.getAsync<T>(serviceIdentifier: ServiceIdentifier<T>): Promise<T>
````
I added the return types for the get methods.

The new methods runbindAsync, unbindAllAsync and rebindAsync have not been document in this wiki page but then neither have **bind** and unbind.  Yet rebind is !

I'd appreciate someone else updating the docs as I am maxed out !

notaphplover commented on test/container/module_activations_store.test.ts:
```diff
@@ -0,0 +1,73 @@
+    }
+    expect(cloneModule2Handlers).to.deep.equal(expectedModule2Handlers)
+  })
+})
```
line feed

tonyhallett replied:
will do

notaphplover commented on src/container/container.ts:
```diff
@@ -27,6 +28,7 @@ class Container implements interfaces.Container {
     private _snapshots: interfaces.ContainerSnapshot[];
     private _metadataReader: interfaces.MetadataReader;
     private _appliedMiddleware: interfaces.Middleware[] = [];
+    private _moduleActivationStore: interfaces.ModuleActivationsStore
```
Maybe it should be named `moduleActivationStore`

tonyhallett replied:
will change

notaphplover reviewed (commented):
Nice! Everything seems to be all right, I just added some minor suggestions

notaphplover committed:
Merge branch 'master' of https://github.com/inversify/InversifyJS into async-v2

tonyhallett committed:
Merge branch 'master' into async-v2

tonyhallett committed:
review changes - type name change and missing snapshot moduleActivationStore test

tonyhallett committed:
last code climate refactor

notaphplover commented on src/container/container.ts:
```diff
@@ -408,11 +408,14 @@ class Container implements interfaces.Container {
+        const moduleHandlers = handlers[activations ? "onActivations":"onDeactivations"];
+        if(moduleHandlers.length > 0) {
+            this[activations ? "_activations":"_deactivations"].removeByCondition(handler => moduleHandlers.indexOf(handler as any) !== -1);
         }
```
We should not refactor this. It's a false positive. Even if code climate suggests a refactor there is no way to improve the code

PodaruDragos replied:
index signatures are always hard to type properly. a ts PR is happening where you can you use a combined index signature based on symbols or even keys of enums. just keep those things in mind, especially me and @notaphplover since we tried to move to stricter rules and eslint.

PodaruDragos replied:
typecasting this things won't do much in this kind of cases since we probably need to cast to unknown first

tonyhallett replied:
If you are not happy with some of these refactors can I leave it for you to revert them ?

notaphplover replied:
Sure, allow me to handle it :). It's still a great job @tonyhallett, thank you so much!

tonyhallett replied:
Ok thanks.

notaphplover commented on src/container/container.ts:
```diff
@@ -494,19 +494,16 @@ class Container implements interfaces.Container {
+                    const bindingToSyntax = this[bindOrRebind](serviceIdentifier);
+                    (bindingToSyntax as any)._binding.moduleId = moduleId;
+                    return bindingToSyntax;
+                };
```
We should not refactor this. It's a false positive. Even if code climate suggests a refactor there is no way to improve the code

PodaruDragos replied:
this is gonna be hard to type :D when we do move to stricter ts rules and eslint

tonyhallett replied:
If you are not happy with some of these refactors can I leave it for you to revert them ?

notaphplover replied:
Sure, allow me to handle it :). It's still a great job @tonyhallett, thank you so much!

PodaruDragos replied:
No no this is fine. I am just writing this so we know when we are gonna do the upgrades. I didn't want to sound rude or anything, just wanted to draw attention to those things

tonyhallett replied:
@PodaruDragos 
No offence taken !

notaphplover committed:
docs: update changelog

notaphplover commented on src/container/container.ts:
```diff
@@ -278,89 +399,221 @@ class Container implements interfaces.Container {
+        }
+        if (handlers.onDeactivations.length > 0) {
+            this._deactivations.removeByCondition(onDeactivation => handlers.onDeactivations.indexOf(onDeactivation) !== -1);
+        }
```
We can improve the performance of this operation, on my way to refactor de activations store

tonyhallett replied:
Sure go ahead.

tonyhallett replied:
@notaphplover 

Instead of putting the module activations in with those added directly from the container, keep them separate and merge them together when required.

e.g have a method on the container instead of below
```
//resolver.ts **** container.getActivationsForService(binding.serviceIdentifier)
const _getContainerActivationsForService = <T>(container: interfaces.Container, serviceIdentifier: interfaces.ServiceIdentifier<T>) => {
    // smell accessing _activations, but similar pattern is done in planner.getBindingDictionary()
    const activations = (container as any)._activations as interfaces.Lookup<interfaces.BindingActivation<any>>;

    return activations.hasKey(serviceIdentifier) ? activations.get(serviceIdentifier).values() : [].values();
}
//container.ts
private _deactivate<T>(binding: Binding<T>, instance: T): void | Promise<void> {
        const constructor = Object.getPrototypeOf(instance).constructor;

        try { //**** this._getDeactivationsForService(binding.serviceIdentifier)
            if (this._deactivations.hasKey(binding.serviceIdentifier)) {
                const result = this._deactivateContainer(
                    instance,
                    this._deactivations.get(binding.serviceIdentifier).values(),
                );
```

When unload no need for _removeModuleHandlers logic - just remove the module bindings from the moduleActivationStore
```
public unload(...modules: interfaces.ContainerModule[]): void {
        modules.forEach((module) => {
            const deactivations = this._removeModuleBindings(module.id)
            this._deactivateSingletons(deactivations);

            this._removeModuleHandlers(module.id);
        });

    }

```

Make getActivationsForService(binding.serviceIdentifier) an interface method and remove the smell

tonyhallett replied:
Although this does not retain activation order

notaphplover committed:
feat: add lookup.removeIntersection

tonyhallett commented on src/container/lookup.ts:
```diff
@@ -60,27 +60,42 @@ class Lookup<T> implements interfaces.Lookup<T> {
         }
+    }
+
+    public removeIntersection(lookup: interfaces.Lookup<T>): void {
```
There is no test for removeIntersection

notaphplover replied:
I'll add it

tonyhallett commented on test/container/module_activation_store.test.ts:
```diff
@@ -5,69 +5,83 @@ import { interfaces } from "../../src/inversify"
 describe("ModuleActivationStore", () => {
   it("should remove handlers added by the module", () => {
```
being picky - now that the serviceIdentifier is used perhaps these tests should have handlers for different service identifiers

notaphplover replied:
Right, I'll add it

tonyhallett reviewed (commented):
Aside from these LGTM

notaphplover committed:
test: add test for lookup.removeIntersection

tonyhallett commented on test/container/lookup.test.ts:
```diff
@@ -177,4 +177,45 @@ describe("Lookup", () => {
+
+    const serviceIdentifier3 = 'service-identifier-3';
+
+    const lookulToIntersectServiceIdentifier3Values = [31, 32, 33, 34];
```
lookul to lookup

tonyhallett reviewed (commented):
Also noticed whilst looking at missing coverage:

On my machine there is one uncovered line in container
`throw new Error(`${ERROR_MSGS.CONTAINER_OPTIONS_INVALID_SKIP_BASE_CHECK}`);`
When this is covered two new lines are reported !

options.defaultScope undefined is not covered
```
public bind<T>(serviceIdentifier: interfaces.ServiceIdentifier<T>): interfaces.BindingToSyntax<T> {
        const scope = this.options.defaultScope || BindingScopeEnum.Transient;
```

The second is important for this pull request.  That there is an error when the service identifier is not in the binding dictionary.
```
public async unbindAsync(serviceIdentifier: interfaces.ServiceIdentifier<any>): Promise<void> {
        if (this._bindingDictionary.hasKey(serviceIdentifier)) {
```

notaphplover committed:
test: improve module_activation_store tests

notaphplover committed:
test: add test to ensure unbind async throws an error when no binding is found

notaphplover reviewed (approved):
LGTM! :)

tonyhallett reviewed (commented):
Great.

More missing coverage if you want to add now otherwise can be done later
container.test.ts
```
it("Should default binding scope to Transient if no default scope on options", () => { 
const container = new Container(); 
container.options.defaultScope = undefined; 
const expectedScope:interfaces.BindingScope = "Transient" 
expect((container.bind("SID") as any)._binding.scope).to.equal(expectedScope) });
```

there is not full test coverage for `if(context.plan)`.  I think the condition can go.
```
function plan(
    metadataReader: interfaces.MetadataReader,
    container: interfaces.Container,
    isMultiInject: boolean,
    targetType: interfaces.TargetType,
    serviceIdentifier: interfaces.ServiceIdentifier<any>,
    key?: string | number | symbol,
    value?: any,
    avoidConstraints = false
): interfaces.Context {

    const context = new Context(container);
    const target = _createTarget(isMultiInject, targetType, serviceIdentifier, "", key, value);

    try {
        _createSubRequests(metadataReader, avoidConstraints, serviceIdentifier, context, null, target);
        return context;
    } catch (error) {
        if (
            isStackOverflowExeption(error)
        ) {
            if (context.plan) {
                circularDependencyToException(context.plan.rootRequest);
            }
        }
        throw error;
    }

}
```

Source code changes LGTM !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Still need the wiki changes mentioned before and dodgy test in resolver.test.ts
`it("Should only call parent async singleton once within child containers"`

tonyhallett commented on test/resolution/resolver.test.ts:
```diff
@@ -1017,170 +1117,1438 @@ describe("Resolve", () => {
+      expect(object1).not.equals(object2);
+  });
+
+  it("Should only call parent async singleton once within child containers", async () => {
```
no parent / child

tonyhallett commented on wiki/container_api.md:
```diff
@@ -334,3 +454,32 @@ expect(ninja.fight()).to.eql("cut!");
+
+## container.onActivation<T>(serviceIdentifier: ServiceIdentifier<T>, onActivation: BindingActivation<T>)
+
+Adds an activation handler for the dependencies's identifier.
```
Adds an activation handler for all dependencies with the specified identifier.

tonyhallett reviewed (commented):
@notaphplover Are you ok if I finish this all off ? Wiki and tests ? I will leave `if(context.plan)` in case dev has messed with context in middleware and cover instead.

tonyhallett committed:
100% coverage

tonyhallett committed:
correct test to match description

tonyhallett committed:
wiki grammar correction

tonyhallett committed:
make container unload methods explicit that applies equally to async

tonyhallett committed:
type changes

tonyhallett committed:
wiki changes

tonyhallett commented on src/planning/planner.ts:
```diff
@@ -246,9 +246,7 @@ function plan(
-            if (context.plan) {
-                circularDependencyToException(context.plan.rootRequest);
-            }
+            circularDependencyToException(context.plan.rootRequest);
```
I decided that the condition was not necessary given that the planner creates the context and sets the plan and there is no way to intercept it.

tonyhallett replied:
I decided that the condition was not necessary given that the planner creates the context and sets the plan and there is no way to intercept it.

tonyhallett commented on src/interfaces/interfaces.ts:
```diff
@@ -202,12 +202,12 @@ namespace interfaces {
         snapshot(): void;
         restore(): void;
-        createChild(): Container;
+        createChild(containerOptions?: interfaces.ContainerOptions): Container;
```
Matching the interface to the Container class.  Perhaps should not do this...

notaphplover replied:
I think it's better to do this in a new PR, I'm adding to my notes

notaphplover reviewed (commented):
Excellent job! Thank you so much @tonyhallett , I think we should avoid to change one interface but everything else is great! Would you agree if I undo the change ar `.createChild()` ?

notaphplover committed:
update interfaces.Container.createChild to avoid receiving parameters

tonyhallett commented on src/resolution/resolver.ts:
```diff
@@ -79,7 +81,7 @@ const _resolveRequest = (requestScope: interfaces.RequestScope) =>
             binding.activated = true;
         } else if (binding.type === BindingTypeEnum.Constructor) {
-            result = binding.implementationType;
+            result = binding.implementationType as unknown as T;
```
There is a problem here due to the typing
```
public toConstructor<T2>(constructor: interfaces.Newable<T2>): interfaces.BindingWhenOnSyntax<T> {
        this._binding.implementationType = constructor as any;
       ...
    }
```

notaphplover replied:
I think we should add breaking changes to solve it. Do you have anything in mind?

We have plans to change typings, but I think we should accept they are not perfect :(

tonyhallett replied:
> Do you have anything in mind?

It will need some thought for sure !

> I think we should add breaking changes to solve it

That will probably be necessary.

It's only a problem as resolveRequest is typed to return `undefined | T | Promise<T> | (T | Promise<T>`
All other returns are typed as any !

Perhaps better to untype the return for now.

Problems still remain but they were there before this feature.

I will happily look at the typings at a later date.

What I would like to do is to remove all of this code !

```
if (binding.type === BindingTypeEnum.ConstantValue) {
            result = binding.cache;
            binding.activated = true;
        } else if (binding.type === BindingTypeEnum.Function) {
            result = binding.cache;
            binding.activated = true;
        } else if (binding.type === BindingTypeEnum.Constructor) {
            result = binding.implementationType as unknown as T;
        } else if (binding.type === BindingTypeEnum.DynamicValue && binding.dynamicValue !== null) {
            result = invokeFactory(
                "toDynamicValue",
                binding.serviceIdentifier,
                () => (binding.dynamicValue as (context: interfaces.Context) => any)(request.parentContext)
            );
        } else if (binding.type === BindingTypeEnum.Factory && binding.factory !== null) {
            result = invokeFactory(
                "toFactory",
                binding.serviceIdentifier,
                () => (binding.factory as interfaces.FactoryCreator<any>)(request.parentContext)
            );
        } else if (binding.type === BindingTypeEnum.Provider && binding.provider !== null) {
            result = invokeFactory(
                "toProvider",
                binding.serviceIdentifier,
                () => (binding.provider as interfaces.Provider<any>)(request.parentContext)
            );
        } else if (binding.type === BindingTypeEnum.Instance && binding.implementationType !== null) {
            result = resolveInstance(
                binding,
                binding.implementationType,
                childRequests,
                _resolveRequest(requestScope)
            );
        } else {
            // The user probably created a binding but didn't finish it
            // e.g. container.bind<T>("Something"); missing BindingToSyntax
            const serviceIdentifier = getServiceIdentifierAsString(request.serviceIdentifier);
            throw new Error(`${ERROR_MSGS.INVALID_BINDING_TYPE} ${serviceIdentifier}`);
        }
```
and instead `binding.provideValue(context)` and have binding as abstract  with a derivation for each of the BindingToSyntax distinguishable methods. 
( requestScope not necessary with the scope feature I want to add when this goes through )

Perhaps this would help.

notaphplover replied:
Well T is supposed to be the result's type, I understand your feeling, I really want to improve that code's block, but (T | Promise<T> | (T|Promise<T>)[]) is a valid returning type. A constructor binding is suposed to be called with a `T extends Newable<unknown>`, considering that  (T | Promise<T> | (T|Promise<T>)[]) is a valid return type. 

I'm looking forward to improve that code, but I believe achieving changes through small PR is the best way to go. It's easier to revier a small PR, it's easier to solve conflicts so it helps this project to scale. So I prefer to solve this in another PR

tonyhallett replied:
We can discuss the return type at a later date.  I agree another PR for type changes.

tonyhallett reviewed (commented):
@notaphplover 
We should really sort this out.

tonyhallett commented on src/interfaces/interfaces.ts:
```diff
@@ -42,19 +48,24 @@ namespace interfaces {
-        onActivation: ((context: interfaces.Context, injectable: T) => T) | null;
+        onActivation: BindingActivation<T> | null;
+        onDeactivation: BindingDeactivation<T> | null;
         cache: T | null;
```
cache Should be T | Promise<T>

notaphplover replied:
Ok let's discuss this at #1319

notaphplover replied:
Or in a new issue, maybe a new issue is better

tonyhallett commented on src/resolution/resolver.ts:
```diff
@@ -45,12 +47,12 @@ const _resolveRequest = (requestScope: interfaces.RequestScope) =>
     } else {
 
-        let result: any = null;
+        let result: undefined | T | Promise<T> | (T | Promise<T>)[];
```
This should be `undefined | T | Promise<T>`

I suggest that also refactor to
```
if (request.target.isOptional() && bindings.length === 0) {
            return undefined;
        }

        const binding = bindings[0];
        return resolveBinding(requestScope, request, binding);

```
```
const resolveBinding = <T>(requestScope: interfaces.RequestScope,request: interfaces.Request, binding:interfaces.Binding<T>):
     T | Promise<T> 
```

tonyhallett replied:
Would you like me to make these changes ?

notaphplover replied:
Ok let's discuss this at #1319

notaphplover replied:
I merged the PR, don't worry, allow me to analyze it and create the PR

notaphplover replied:
After analizyng this, result could be a `(T | Promise<T>)[]`:

```ts
        } else if (binding.type === BindingTypeEnum.Instance && binding.implementationType !== null) {
            result = resolveInstance<T>(
                binding,
                binding.implementationType,
                childRequests,
                _resolveRequest<T>(requestScope)
            );
        } else {
```

tonyhallett replied:
resolveInstance creates an instance.  Return type `T | Promise<T>`

What am I missing ?!

notaphplover replied:
My bad, you are right. I'm adding a new issue to solve it.

tonyhallett replied:
Can you extract the resolve binding method as it allows strong typing of some of the return values - e.g binding.cache
```
const resolveBinding = <T>(requestScope: interfaces.RequestScope,request: interfaces.Request, binding:interfaces.Binding<T>):
     T | Promise<T> 

```

notaphplover replied:
I added it to #1321 :)

tonyhallett replied:
👍

tonyhallett reviewed (commented):
Also please see https://github.com/inversify/InversifyJS/issues/1319
